### DISCUSSION TITLE
Convert RGB PNG to Indexed colors

### DISCUSSION THREAD

**Discussion by cartometro** [ImageMagick/ImageMagick#2485] [Help] [answered]
Hello,
I'm trying to convert image to indexed color 
Here is the original image:
![test](https://user-images.githubusercontent.com/70488598/91668906-dee86e00-eb10-11ea-996c-9af1cbf2500b.png)
Identify ; `test.png PNG 300x299 300x299+0+0 8-bit sRGB 5099B 0.000u 0:00.000`

**First try**
Using the command line :
`magick test.png -colors 50 test_magick_color50.png`
I got two main issues:
1. The result image is only 15 colors and, as a result, looks far from the original
![test_magick_palette50](https://user-images.githubusercontent.com/70488598/91669007-c7f64b80-eb11-11ea-8595-b038cb5e954b.png)
Identify ; `test_magick_color50.png PNG 300x299 300x299+0+0 8-bit sRGB 16c 7960B 0.000u 0:00.000`

2. The result image size 9,69KB  is bigger than the original RGB 4,97KB
Using  `-quality 90 -strip +profile *`, I got a 9,17KB image, still much bigger !

**Second try**
Using 
`magick test.png -type Palette -strip test_magick_palette.png`
The result is a 60 colors PNG (I guess that on more colorfull images, I'll get 200 or more colors...) and filesize 7,63K. Better but still bigger than original !
![test_magick_palette_q8](https://user-images.githubusercontent.com/70488598/91669326-58ce2680-eb14-11ea-9add-2f69bac2fb12.png)
Identify ; `test_magick_palette.png PNG 300x299 300x299+0+0 8-bit sRGB 60c 8033B 0.000u 0:00.000`

In comparison, using GIMP, the image is only 2,21KB and has exactly 50 colors, with a good looking result:
![test_gimp_palette50](https://user-images.githubusercontent.com/70488598/91669030-f83dea00-eb11-11ea-9278-62922784ab71.png)
Identify ; `test_gimp_palette50.png PNG 300x299 300x299+0+0 8-bit sRGB 50c 2273B 0.000u 0:00.000`

How could I force an exact number of color and reduce filesize ? (even without reaching the GIMP size, but smaller that RGB...)

Thanks a lot for your help,

Using version: `ImageMagick 7.0.10-28 Q8 x64 2020-08-16` (NB: results are the same with Q16 as image is 8bits)

**Comment by snibgo** [comment]
The image has 172 unique colours.
```
magick 91668906-dee86e00-eb10-11ea-996c-9af1cbf2500b.png -type palette x.png
```
This reduces the size from 170121 bytes to 3973 bytes, without changing the image.

IM does not have an operation that will reduce the image to _exactly_ a number you specify.

**Comment by cartometro** [comment]
Thanks for the reply,

Interresting, which version of IM do you use ? I don't get the same result  :
```
magick 91668906-dee86e00-eb10-11ea-996c-9af1cbf2500b.png -verbose -type palette x.png

91668906-dee86e00-eb10-11ea-996c-9af1cbf2500b.png=>x.png PNG 300x299 300x299+0+0 8-bit Palette sRGB 172c 5099B 0.094u 0:00.119
```

The result file size is 8033B on disk

**Reply by fmw42** [reply]
-kmeans will reduce the colors to a specified number. But it is slow. So do not specify a lot of colors. Also you can specify a color map (image) and use -remap to remap to those colors.

**Reply by cartometro** [reply]
Thanks to reveal me that option ! There are not many references, probably because it's quite new ([link](https://www.imagemagick.org/discourse-server/viewtopic.php?t=37301)).
The reduction is not exactly 50 colors on the sample (only 25) but the colormap selection is very good !
![z10](https://user-images.githubusercontent.com/70488598/91764466-a5c70100-ebd7-11ea-8b6b-2f9628447959.png)
I'll benchmark that function on more files, and maybe tweek it as is is very configurable !

(The option to use an existing colormap is not applicable, I'm processing tiles which could be very differents....)

**Comment by snibgo** [comment]
That was with v7.0.8-64.

With v7.0.10-27, which I don't normally use, the image is reduced to 65 colours (so the image is changed), and the size is 13224 bytes. But if I add "+dither", the colours are not reduced, and the image is not changed, and the size is 3973 bytes which is the same as with v7.0.8-64.

This is weird. I suspect a bug in v7.0.10-27.

Try adding "+dither" to your command:
```
magick 91668906-dee86e00-eb10-11ea-996c-9af1cbf2500b.png -verbose +dither -type palette x.png
```

**Comment by cartometro** [comment]
Indeed, with the dither command the result file is much smaller on this sample ! 
I've applied the manipulation to a batch of 70.000+ files and there is not much différences on the overall : 408MB vs 401MB with the +dither option (but the processing time seems much faster with the +dither option, I need more benchmark to be sure).

This is to compare to the only 130MB of Gimp applying and exact color reduction to 50 colors (and 8 additional hours of processing...)

I got a last issue with the `+dither -type Palette` command, a transparent background is dropped :(

I'll try fmw42 suggestion on my batch of file to see the result. It result look interresting, but have to benchmark the time !

**Reply by fmw42** [reply]
Try saving as PNG8:output.png. Also you can try changing the -quality and compression. See https://imagemagick.org/Usage/formats/#png_write. Also you can post process with one of the external PNG tools that compress better, such as pngcrush. See https://imagemagick.org/Usage/formats/#png_non-im

**Comment by cartometro** [comment, marked as answer]
After some tests, the [-kmeans option](https://www.imagemagick.org/discourse-server/viewtopic.php?t=37301) reveals very good results on small tiles images (300x300) with 50 colors. The color selection is good and the resulting filesize smaller than with Gimp ! 
I've also tried the set the iterations option to 50 (instead of the default 100) but I didn't changed the processing time (In fact it takes slightly more time, a bit strange to my understanding of the iteration option...).

For reference, the previous processing to generate 70.000+ tiles from a 16 master tiles (18.000x18.000 pixels) takes:
- 2h40 with -colors 50 option, with bad results, and 400+ MB filesize
- 10h+ with Gimp, also reducing tiles to 50 colors, and 128MB filesize
- 1h50 with -kmeans 50 option for a total of 86MB

``` magick convert source.png  -kmeans 50  -quality 90  -strip +profile * export.png ```

So, no doubt that I'll use the -kmeans option ! Thanks-your @fmw42 for your this work !

For big images (10.000x10.000) which required more colors (150) the kmeans algorithm is indeed very slow (but result is better than with -colors option).  I'll have a look at pngcrush to optimize theses big exports (not really needed on small tiles images)...

(sorry fmw42, I cant mark your reply as an answer as it's not a standalone reply...)

**Reply by fmw42** [reply]
You can speed up -kmeans, if you reduce the number of iterations or increase the convergence threshold.
